parley/src/analysis/cluster.rs

```diff
@@ -0,0 +1,425 @@
+
+                // Copy the characters back to our form structure
+                let mut i = 0;
+                for c in nfd_str.chars() {
+                    if i == MAX_CLUSTER_SIZE {
+                        return None;
+                    }
+
+                    // Use the first character as a template for other properties
+                    let mut copy = self.chars[0];
+                    copy.ch = c;
+                    if i >= self.decomp.chars.len() {
+                        self.decomp.chars.push(copy);
+                    } else {
+                        self.decomp.chars[i] = copy;
+                    }
+                    i += 1;
+                }
+
+                if i == 0 {
+                    return None;
+                }
+
+                self.decomp.len = i as u8;
+                self.decomp.state = FormState::Valid;
+                self.decomp.setup();
+                Some(self.decomp.chars())
+            }
+            FormState::Valid => Some(self.decomp.chars()),
+        }
+    }
+
+    fn composed(
+        &mut self,
+        analysis_data_sources: &AnalysisDataSources,
+        scratch_string: &mut String,
+    ) -> Option<&[Char]> {
+        match self.comp.state {
+            FormState::Invalid => None,
+            FormState::None => {
+                // First, we need decomposed characters
+                if self
+                    .decomposed(analysis_data_sources, scratch_string)
+                    .map(|chars| chars.len())
+                    .unwrap_or(0)
+                    == 0
+                {
+                    self.comp.state = FormState::Invalid;
+                    return None;
+                }
+
+                self.comp.state = FormState::Invalid;
+
+                // Create a string from the decomposed characters to normalize
+                scratch_string.clear();
+                for ch in &self.decomp.chars()[..self.decomp.len as usize] {
+                    scratch_string.push(ch.ch);
+                }
+
+                // Normalize to NFC (composed) form
+                let nfc_str = analysis_data_sources
+                    .composing_normalizer()
+                    .normalize(scratch_string);
+
+                // Copy the characters back to our form structure
+                let mut i = 0;
+                for c in nfc_str.chars() {
+                    if i >= MAX_CLUSTER_SIZE {
+                        self.comp.state = FormState::Invalid;
+                        return None;
+                    }
+
+                    // Use the first decomposed character as a template for other properties
+                    let mut ch_copy = self.decomp.chars[0];
+                    ch_copy.ch = c;
+                    if i >= self.comp.chars.len() {
+                        self.comp.chars.push(ch_copy);
+                    } else {
+                        self.comp.chars[i] = ch_copy;
+                    }
+                    i += 1;
```

Comment:
Same as above. Just using pairwise [compose](https://docs.rs/icu/latest/icu/normalizer/properties/struct.CanonicalCompositionBorrowed.html#method.compose) should be sufficient. We can improve this or possibly use shaper driven itemization later (#462).